docs/plutus-ir/plutus-ir.md

```diff
@@ -0,0 +1,193 @@
+# Plutus IR design proposal
+
+## Syntax
+
+The syntax is mostly as in Plutus Core. Where we modify the syntax of Plutus Core we will
+list only the additions and removals, marking them explicitly with `+` or `-`.
+We also require syntax with lists of subparts. We denote a list of `t`-productions with `t..`.
+
+```
+VarDecl VD      := (vardecl x T)
+TyVarDecl TVD   := (tyvardecl alpha K)
+
+Type R, S, T    := - (fix alpha T)
+Term L, M, N    := + (let BK B.. M)
+
+BiningKind BK   := (nonrec)
+                   (rec)
+Binding B       := (termbind VD L)
+                   (typebind TVD T)
+                   (datatypebind D)
+
+Datatype D      := (datatype
+                      TVD   # the type variable for the datatype
+                      TVD.. # the argument type variables of the datatype
+                      x     # the name of the destructor function
+                      C..   # the constructors of the datatype
+                   )
+Constructor C   := (constr
+                       x   # the name of the constructor
+                       T.. # the argument types of the constructor
+                    )
+```
+
+Here is an example, constructing `[1]`:
+```
+# bind List
+(let (rec)
+    (datatypebind
+        (datatype
+            (tyvardecl List (fun (type) (type)))
+            (tyvardecl a (type))
+            match_List
+            (constr
+                Nil
+            )
+            (constr
+                Cons
+                a
+                [List a]
+            )
+        )
+    )
+    # bind int64 for convenience
+    (typebind (tyvardecl int (type)) [(con integer) (con 64)])
+    # apply cons to 1 and nil
+    [ { Cons int } (con 64 ! 1) { Nil int } ]
```

Comment:
This looks weird. If we elaborate the iterated application, we'll get

```
[ [ { Cons int } (con 64 ! 1) ] { Nil int } ]
```

Do we collapse `[{ ... }]` into `{...}`?

For me the Agda-like syntax is much more readable:

```
Cons {int} (con 64 ! 1) (Nil {int})
```

Reply:
I don't see what's weird about it? You have to instantiate `Cons` to the type parameter, then you apply it to the head and the tail. The tail is `Nil`, which also needs to be instantiated.